x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -484,6 +509,247 @@ protected LogicalPlan rule(OrderBy ob) {
+                Set<Expression> toBreakup = newSetFromMap(new IdentityHashMap<>());
+                condition.forEachUp(e -> {
+                    Markers markers = markerMap.get(e);
+                    // the reason of this limitation is https://github.com/elastic/elasticsearch/issues/23874
+                    // AND can be broken up into two pieces, but we cannot break up anything else
+                    if (markers.hasGrouping && markers.hasAggregate && e instanceof And) {
+                        breakupCandidates.add(e);
+                        // only add, if the issues did not propagate from the children
+                        if (e.children().stream().noneMatch(breakupCandidates::contains)) {
+                            toBreakup.add(e);
+                        }
+                    }
+                });
+                
+                // let's break up the condition into two pieces
+                conditionOnAggregates = condition.transformUp(e -> {
+                    if (toBreakup.contains(e)) {
+                        And and = (And) e;
+                        // one of the children is a condition on grouping, let's remove that
+                        // and return the other item (And will always have two children)
+                        List<Expression> newAndChildren = and.children().stream().map(andExp -> {
+                            if (markerMap.get(andExp).hasGrouping) {
+                                conditionsOnGroupings.add(andExp);
+                                return Literal.TRUE;
+                            } else {
+                                return andExp;
+                            }
+                        }).collect(Collectors.toList());
+                        return and.replaceChildrenSameSize(newAndChildren);
+                    }
+                    return e;
+                });
+            } else {
+                // has only conditions on groupings (or with literals) -> the whole condition can be pushed down
+                conditionsOnGroupings.add(condition);
+            }
+            
+            if (conditionsOnGroupings.isEmpty()) {
+                return plan;
+            }
+            
+            Expression conditionOnGroupings = Predicates.combineAnd(conditionsOnGroupings);
+            LogicalPlan newChild = plan.child().transformUp(p -> {
+                if (p == aggregate) {
+                    // push down the filters on the groupings below the Aggregate
+                    Filter filterOnGroupings = new Filter(Source.EMPTY, aggregate.child(), conditionOnGroupings);
+                    return new Aggregate(aggregate.source(), filterOnGroupings, aggregate.groupings(), aggregate.aggregates());
+                }
+                return p;
+            });
+            
+            return conditionOnAggregates == null ? newChild : new Filter(plan.source(), newChild, conditionOnAggregates);
+        }
+
+        public static void verify(
```

Comment:
This function seems only to only be called from the Verifier: would it make sense to have it here?